File: leetcode001.cpp
//============================================================================
// Name        : leetcode001.cpp
// Author      : 
// Version     :
// Copyright   : Your copyright notice
// Description : Hello World in C++, Ansi-style
//============================================================================

#include <iostream>
#include <vector>
#include <algorithm>
#include <stddef.h>
using namespace std;

struct Node {         //define a struct including index and value of a vector
	int index;
	int value;
	Node(int i, int v): index(i), value(v){} //define the elements of a struct Node
};

bool compare(const Node &left, const Node &right) {   //compare if the vector is set up from small to large
	return left.value < right.value;
}

class Solution {
public:
	vector<int> twoSum(vector<int> &numbers, int target) {
		vector <Node> a;
		for(int i=0; i<numbers.size(); i++) {
			a.push_back(Node(i+1, numbers[i]));
		}
		std::sort(a.begin(), a.end(), compare);
		int beg = 0;
		int end = a.size() - 1;

		while (beg < end) {
			int sum = a[beg].value + a[end].value;
			if (sum==target) {
				int small = std::min(a[beg].index, a[end].index);
				int big = std::max(a[beg].index, a[end].index);
				vector<int> ret;
				ret.push_back(small);
				ret.push_back(big);
				return ret;
			}
			else if(sum < target) {
				beg++;
			}
			else {
				end--;
			}
		}
	}
};

int main()
{
	int da[5] = {2,5,7,11,15};
	vector<int> data;
	int i=0;
	while(i<5)
	{
		data.push_back(da[i]);
		i++;
	}
	vector<int> result;
	int target = 9;
	Solution test;
	result = test.twoSum(data, target);
	cout<<data.size()<<endl;
	cout<<result.at(0)<<endl<<result.at(1)<<endl;

	int *p = NULL;
	if(p)
		cout<<p;
}

File: Leetcode001.java
package test;

public class Leetcode001 {
	public int[] towSum(Node[] num, int target) {
		int[] index = new int[2];
		int i = 0;
		int j = num.length - 1;
		while(num[i].number + num[j].number != target) {
			if(num[i].number + num[j].number > target)
				j--;
			else
				i++;
		}
		index[0] = num[i].index;
		index[1] = num[j].index;
		return index;
	}
	
	void quicksort(Node[] num, int start, int end) {
		if (start < end) {
			int q = partition(num, start, end);
			quicksort(num, start, q-1);
			quicksort(num, q+1, end);
		}
	}
	
	int partition(Node[] num, int start, int end) {
		Node temp;
		int x = num[end].number;
		int i = start - 1;
		for(int j = start; j < end; j++) {
			if (num[j].number <= x) {
				i++;
				temp = num[i];
				num[i] = num[j];
				num[j] = temp;
			}
		}
		temp = num[i+1];
		num[i+1] = num[end];
		num[end] = temp;
		return i+1;
	}
	
	public static void main(String args[]) {
		int array[] = {2, 11, 7, 15};
		int target = 9;
		Node nod[] = new Node[array.length];
		for(int i=0; i<array.length; i++)
			nod[i] = new Node(i+1, array[i]);
		
		Leetcode001 test = new Leetcode001();
		test.quicksort(nod, 0, array.length - 1);
		
		int[] index = new int[2];
		index = test.towSum(nod, target);
		System.out.println("index1 = " + index[0] + "\n" + "index2 = " + index[1]);
	}
}

class Node {
	public int index;
	public int number;
	
	Node(int ind, int num) {
		this.index = ind;
		this.number = num;
	}
}
